### PR TITLE
Add option to validate input data both in CLI and Renderer 

### PR DESCRIPTION
**Pull request by mszostok**
## Description

Changes proposed in this pull request:

- Add `TypeRef` for `InputParameter` in GraphQL schema: needs to be continued in https://github.com/capactio/capact/issues/426
   - Current implementation is tested in unit-test but cannot be tested in real scenario.

- Known issue: The survey library doesn't clear the last error even if valid input was provided.
   Screenshot 2021-08-09 at 17 21 49

- Renderer part:
    - Validate input data before render process starts.
    - Create dedicated methods to show our known issue for input parameters, generalize usage.
    - Pass loaded input data into renderer. As a result, we don't need to fetch it once again from K8s secret.
    - Add validator adapter for workflow.

- CLI:
	- Optimize `FindInterfaceRevision` query to fetch only needed fields.
	- Use validator both for loaded files and in interactive mode.

- Validator: 
    - Add opt to load input parameters and TypeInstances
        - It could be extracted into sth like `loader` pkg, but for now it’s much simpler when it’s under single instance.
    - Sort validation result by name. As a result, we have a deterministic error message.
    - Add result validation builder. Maybe in future we will be able to use it in other places too.

## Question:

Q: Should we allow unknown TypeInstance for `capact act create`? 
A: Currently, we allow that, so we can pass `additionalTypeInstances` which are not specified on Interface e.g. existing database. But we should once again recheck how we want to handle `additionalTypeInstances`. Currently, the implemented policy doesn't allow us to do it, as it works only with `requires` property. I see inconsistency between injected `additionalParameters` and `additionalTypeInstaces`. The other option is to support that during advanced rendering.

In code, all places that need to be changed are marked with `TODO(advanced-rendering/policy):` comment.

## Testing

1. Create cluster with populated Public Hub: `make dev-cluster`
2. Checkout PR and build CLI: `make build-tool-cli`
3. Create example TypeInstance:
   ```bash
   cat > /tmp/valid-ti.yaml << ENDOFFILE
   typeInstances:
     - typeRef:
         path: cap.type.capactio.capact.validation.update
         revision: "0.1.0"
       value:
         key: true
   ENDOFFILE

   export VALID_TI_ID=$(capact ti create -f /tmp/valid-ti.yaml -ojson | jq -r '.[].id')
   ```

   ```bash
   cat > /tmp/not-valid-ti.yaml << ENDOFFILE
   typeInstances:
     - typeRef:
         path: cap.type.capactio.capact.validation.failing
         revision: "0.1.0"
       value:
         key: true
   ENDOFFILE

   export NOT_VALID_TI_ID=$(capact ti create -f /tmp/not-valid-ti.yaml -ojson | jq -r '.[].id')
   ```

**Capact CLI:**
1. Not existing Interface:
   ```bash
   capact act create cap.interface.capactio.capact.knok.knok.whos.there.nobody
   ```

   Output:
    ```
    Error: Interface cap.interface.capactio.capact.knok.knok.whos.there.nobody was not found in Hub
    ```
1. Missing parameters and TypeInstances
    ```bash
    capact act create cap.interface.capactio.capact.validation.action.update
    ```

    Output:
    ```bash
    Error: 2 validation errors detected:
    - TypeInstances "testUpdate":
        * required but missing TypeInstance of type cap.type.capactio.capact.validation.update:0.1.0
    - Parameters "input-parameters":
        * required but missing input parameters for Interface
    ```
1. Not existing TypeInstance
    ```bash
    cat > /tmp/params.yaml << ENDOFFILE
    testString: "test"
    ENDOFFILE
    cat > /tmp/ti.yaml << ENDOFFILE
    typeInstances:
    - name: "testUpdate"
      id: "id-not-found"
    ENDOFFILE

    capact act create cap.interface.capactio.capact.validation.action.update --parameters-from-file /tmp/params.yaml --type-instances-from-file /tmp/ti.yaml
    ```
   Output:

    ```bash
    Error: 1 validation error detected:
    - TypeInstances "testUpdate":
        * TypeInstance was not found in Hub
    ```

1. Wrong TypeInstance Type
    ```bash
    cat > /tmp/params.yaml << ENDOFFILE
    testString: "test"
    ENDOFFILE
    cat > /tmp/ti.yaml << ENDOFFILE
    typeInstances:
    - name: "testUpdate"
      id: "${NOT_VALID_TI_ID}"
    ENDOFFILE

    capact act create cap.interface.capactio.capact.validation.action.update --parameters-from-file /tmp/params.yaml --type-instances-from-file /tmp/ti.yaml
    ```
   
   Output:
    
    ```bash
    Error: 1 validation error detected:
    - TypeInstances "testUpdate":
        * must be of Type "cap.type.capactio.capact.validation.update" but it's "cap.type.capactio.capact.validation.failing"
    ```

1. Wrong parameters
    ```bash
    cat > /tmp/params.yaml << ENDOFFILE
    key: "test"
    ENDOFFILE
    cat > /tmp/ti.yaml << ENDOFFILE
    typeInstances:
    - name: "testUpdate"
      id: "${VALID_TI_ID}"
    ENDOFFILE

    capact act create cap.interface.capactio.capact.validation.action.update --parameters-from-file /tmp/params.yaml --type-instances-from-file /tmp/ti.yaml
    ```
   
   Output:

    ```
    Error: 1 validation error detected:
    - Parameters "input-parameters":
        * (root): testString is required
    ```   
1. Ignore unknown TypeInstance
    ```bash
    cat > /tmp/params.yaml << ENDOFFILE
    testString: "test"
    ENDOFFILE
    cat > /tmp/ti.yaml << ENDOFFILE
    typeInstances:
    - name: "testUpdate"
      id: "${VALID_TI_ID}"
    - name: "implSpecific"
      id: "${NOT_VALID_TI_ID}"
    ENDOFFILE

    capact act create cap.interface.capactio.capact.validation.action.update --parameters-from-file /tmp/params.yaml --type-instances-from-file /tmp/ti.yaml
    ```

1. Play with interactive mode
   Features:
   1. Validates input and TypeInstance after saving editor
   2. Don't ask if input params are not required
   3. Fill form with TypeInstances' names and add comment with expected Type
   ```bash
   capact act create --name test cap.interface.capactio.capact.validation.action.update -i
   ```

**Engine:**

Repeat steps from 1 to 5 but specify `--validate=false` for each command. This will send request without validation. Check the error message on Action CR.

Additional validation:

1. Wrong additional parameters
    
    Update `cap.implementation.capactio.capact.validation.action.update` and add:
    ```yaml
      additionalInput:
        parameters:
          typeRef:
            path: cap.type.capactio.capact.validation.download
            revision: 0.1.0
    
    ```
    or populate Hub from this branch: https://github.com/capactio/hub-manifests/compare/main...mszostok:test-engine-validate?expand=1, via `kubectl set env deploy/capact-hub-public -n capact-system -c="hub-public-populator" MANIFESTS_PATH="github.com/mszostok/hub-manifests?ref=test-engine-validate"`

    ```bash
    cat > /tmp/params.yaml << ENDOFFILE
    testString: "test"
    ENDOFFILE
    
    cat > /tmp/policy.yaml << ENDOFFILE
    rules:
      - interface:
          path: cap.interface.capactio.capact.validation.action.update
        oneOf:
          - implementationConstraints:
              path: "cap.implementation.capactio.capact.validation.action.update"
            inject:
              additionalInput:
                additional-parameters:
                  ingress:
                    enabled: false
    ENDOFFILE
    
    cat > /tmp/ti.yaml << ENDOFFILE
    typeInstances:
    - name: "testUpdate"
      id: "${VALID_TI_ID}"
    ENDOFFILE
    
    capact act create --name test cap.interface.capactio.capact.validation.action.update --parameters-from-file /tmp/params.yaml --type-instances-from-file /tmp/ti.yaml --action-policy-from-file /tmp/policy.yaml --validate=false
    ```

1. Wrong input additional TypeInstances
    ```bash
    cat > /tmp/params.yaml << ENDOFFILE
    testString: "test"
    ENDOFFILE
    cat > /tmp/ti.yaml << ENDOFFILE
    typeInstances:
    - name: "testUpdate"
      id: "${VALID_TI_ID}"
    - name: "secondKey"
      id: "${NOT_VALID_TI_ID}"
    ENDOFFILE
    
    capact act create cap.interface.capactio.capact.validation.action.update --parameters-from-file /tmp/params.yaml --type-instances-from-file /tmp/ti.yaml
    ```

## Related issue(s)

- https://github.com/capactio/capact/issues/425

### CODE REVIEW

File: internal/cli/action/create_opts.go [hunk 1]
```diff
@@ -139,61 +199,113 @@ func (c *CreateOptions) resolveFromFiles() error {
...
-	if err := survey.AskOne(askAboutTI, &provideTI); err != nil {
-		return nil, err
+func (c *CreateOptions) askForInputTypeInstances() ([]types.InputTypeInstanceRef, error) {
+	body, requiredTI := c.getTypeInstancesForEditor()
+
+	// If input is not required, still ask user whether he wants to specify one.
+	// REASON: we don't support advancedRendering so as a workaround we pass them directly
+	// to the created Action.
+	// TODO(advanced-rendering): remove me.
+	// TODO: or action policy already solves this problem?
```
Comment (mszostok) [inline internal/cli/action/create_opts.go:240]: This needs to be discussed, currently I don't see other option.

File: pkg/sdk/renderer/argo/renderer.go [hunk 2]
```diff
@@ -116,36 +126,51 @@ func (r *Renderer) Render(ctx context.Context, input *RenderInput) (*RenderOutpu
 
 	// 5. List data based on policy and inject them if provided
 	// 5.1 TypeInstances
-	typeInstancesToInject := policyEnforcedClient.ListTypeInstancesToInjectBasedOnPolicy(rule, implementation)
-	err = dedicatedRenderer.InjectDownloadStepForTypeInstancesIfProvided(rootWorkflow, typeInstancesToInject)
+	additionalTypeInstances := policyEnforcedClient.ListTypeInstancesToInjectBasedOnPolicy(rule, implementation)
```
Comment (mszostok) [inline pkg/sdk/renderer/argo/renderer.go:129]: This works only with `requires`, so it's a "fake" additionalTypeInstances. Needs to be discussed.

**Comment by mszostok** [comment]
FYI: checking why integration tests failed

File: internal/cli/action/create_opts.go [hunk 3]
```diff
@@ -139,61 +196,112 @@ func (c *CreateOptions) resolveFromFiles() error {
...
 		return nil, err
 	}
 
 	return toTypeInstance([]byte(editor))
 }
 
+func (c *CreateOptions) getTypeInstancesForEditor() (body string, required bool) {
+	if len(c.ifaceTypes) == 0 {
+		return heredoc.Doc(`
+						# Interface doesn't specify input TypeInstance.
```
Comment (mszostok) [inline internal/cli/action/create_opts.go:285]: sth with padding, in my editor looks ok, but I will fix it

File: internal/cli/action/helpers.go [hunk 4]
```diff
@@ -26,6 +26,24 @@ func isDNSSubdomain(val interface{}) error {
 	return nil
 }
 
+func validatorAdapter(validate func(inputParams string) error) survey.Validator {
+	return func(val interface{}) error {
+		str, ok := val.(string)
+		if !ok {
+			return fmt.Errorf("Cannot enforce input parameters JSONSchema validation on response of type %T", val)
```
Comment (mszostok) [inline internal/cli/action/helpers.go:33]:
```suggestion
			return fmt.Errorf("Cannot enforce parameters validation on response of type %T", val)
```

File: pkg/sdk/renderer/argo/renderer.go (hunk 2 above)
Comment (mszostok) [reply pkg/sdk/renderer/argo/renderer.go:129]: Discussed on our engineering call. As a result, we created a dedicated issue to fix it: https://github.com/capactio/capact/issues/438

File: internal/cli/action/create.go [hunk 5]
```diff
@@ -53,3 +67,38 @@ func Create(ctx context.Context, opts CreateOptions, w io.Writer) (*CreateOutput
 		Namespace: opts.Namespace,
 	}, nil
 }
+
+func setupValidatorWithOpts(ctx context.Context, opts *CreateOptions, hubCli client.Hub) error {
```
Comment (mszostok) [inline internal/cli/action/create.go:71]:
```suggestion
func setupCreateOptsWithValidator(ctx context.Context, opts *CreateOptions, hubCli client.Hub) error {
```

File: pkg/validate/result_aggregator.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+//   err = rs.Report(validator.ValidateParameters(ctx, ifaceSchemas, params))
+func (r *ValidationResultAggregator) Report(result ValidationResult, err error) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	if err != nil {
+		return err
+	}
+	r.issuesCnt += result.Len()
+	if err := result.ErrorOrNil(); err != nil {
```
Comment (pkosiec) [inline pkg/validate/result_aggregator.go:27]: Isn't it better to just get all error messages from the `ValidationResult`? All you do inside `result.ErrorOrNil()` with these error messages you concatenate them:`errors.New(strings.Join(msgs, "\n"))`. 

Pros of such approach:
- Easier to understand the flow (right now it's too complicated in my opinion)
- No `issuesCnt` needed, you just use `len(issuesMsgs)`
- More control on formatting "sub-errors" from a single `ValidationResult` inside `ValidationResultAggregator`.

File: pkg/validate/types.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
+package validate
```
Comment (pkosiec) [inline pkg/validate/types.go:1]: BTW I think we need to unify the package naming. We have manifest validation inside `manifest` package as far as I remember - maybe we should have Action validation in some `action` one? To discuss

**Review by pkosiec — COMMENTED** [review]
🚀 Awesome work! Few comments with improvements suggestions 🙂

File: pkg/validate/types.go (hunk 7 above)
Comment (mszostok) [reply pkg/validate/types.go:1]: after f2f discussion our approach is:
- `pkg/validation/action`
- `pkg/validation/manifest`
- `pkg/validation/` - common stuff

File: pkg/validate/facade/wf_validator.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
+package facade
```
[the remark this answers is not in the document]
Comment (mszostok) [reply pkg/validate/facade/wf_validator.go:1]: moved to `pkg/renderer` 👍

File: pkg/validate/result_aggregator.go (hunk 6 above)
Comment (mszostok) [reply pkg/validate/result_aggregator.go:27]: discussed f2f currently it's too complex to be flattened.

File: internal/cli/action/create_opts.go [hunk 9]
```diff
@@ -42,9 +55,46 @@ func (c *CreateOptions) SetDefaults() {
...
+		}
+	}
+
+	if c.parameters == nil && c.isInputParamsRequired {
+		bldr := validate.NewResultBuilder("Parameters")
+		bldr.ReportIssue(argo.UserInputName, "required but missing input parameters for Interface")
+		if err := r.Report(bldr.Result(), nil); err != nil {
+			return err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (mszostok) [reply internal/cli/action/create_opts.go:76]: those checks are already inside validator, so it's redundant. I just removed them 👍

**Review by pkosiec — APPROVED** [review]
LGTM 🚀

**Pull request merged**